toolsrc/src/vcpkg/portfileprovider.cpp

```diff
@@ -263,4 +307,110 @@ namespace vcpkg::PortFileProvider
 
         return ret;
     }
+
+    VersionedPortfileProvider::VersionedPortfileProvider(const vcpkg::VcpkgPaths& paths) : paths(paths) { }
+
+    const std::vector<vcpkg::Versions::VersionSpec>& VersionedPortfileProvider::get_port_versions(
+        const std::string& package_spec) const
+    {
+        auto cache_it = versions_cache.find(package_spec);
+        if (cache_it != versions_cache.end())
+        {
+            return cache_it->second;
+        }
+
+        auto maybe_versions_json_path = get_versions_json_path(paths, package_spec);
+        if (!maybe_versions_json_path.has_value())
+        {
+            // TODO: Handle db version not existing
+            Checks::exit_fail(VCPKG_LINE_INFO);
+        }
+        auto versions_json_path = maybe_versions_json_path.value_or_exit(VCPKG_LINE_INFO);
+
+        auto versions_json = Json::parse_file(VCPKG_LINE_INFO, paths.get_filesystem(), versions_json_path);
+
+        // NOTE: A dictionary would be the best way to store this, for now we use a vector
+        if (versions_json.first.is_object())
+        {
+            const auto& versions_object = versions_json.first.object();
+            auto maybe_versions_array = versions_object.get("versions");
+
+            if (maybe_versions_array && maybe_versions_array->is_array())
+            {
+                const auto& versions_array = maybe_versions_array->array();
+                for (const auto& version : versions_array)
+                {
+                    const auto& version_obj = version.object();
+                    Versions::VersionSpec spec = extract_version_spec(package_spec, version_obj);
+
+                    auto& package_versions = versions_cache[spec.package_spec];
+                    package_versions.push_back(spec);
+
+                    auto git_tree = version_obj.get("git-tree")->string().to_string();
+                    git_tree_cache.emplace(std::move(spec), git_tree);
+                }
+            }
+        }
+
+        return versions_cache.at(package_spec);
+    }
+
+    ExpectedS<const SourceControlFileLocation&> VersionedPortfileProvider::get_control_file(
+        const vcpkg::Versions::VersionSpec& version_spec) const
+    {
+        auto cache_it = control_cache.find(version_spec);
+        if (cache_it != control_cache.end())
+        {
+            return cache_it->second;
+        }
+
+        // Pre-populate cache.
+        get_port_versions(version_spec.package_spec);
+
+        auto git_tree_cache_it = git_tree_cache.find(version_spec);
+        if (git_tree_cache_it == git_tree_cache.end())
+        {
+            // TODO: Try to load port from database
+            Checks::exit_fail(VCPKG_LINE_INFO);
+        }
+
+        const std::string git_tree = git_tree_cache_it->second;
+        const auto port_directory = paths.git_checkout_port(version_spec.package_spec, git_tree);
```

Comment:
```suggestion
        auto port_directory = paths.git_checkout_port(version_spec.package_spec, git_tree);
```

This can't be const if you want to move it below on line 385